lisp/ess-r-flymake.el

```diff
@@ -62,24 +64,45 @@ each element is passed as argument to 'lintr::with_defaults'."
 
 (defvar-local ess-r--flymake-proc nil)
 
+(defvar-local ess-r--lintr-file nil
+  "Location of the .lintr file for this buffer.")
+
 (defvar ess-r--flymake-def-linter
   (replace-regexp-in-string
    "[\n\t ]+" " "
-   "esslint <- function(str, linters, cache) {
+   "esslint <- function(str, ...) {
     if (!suppressWarnings(require(lintr, quietly=T))) {
         cat('@@error: @@`lintr` package not installed')
     } else {
         if (packageVersion('lintr') <= '1.0.1') {
             cat('@@error: @@Need `lintr` version > v1.0.1')
         } else {
-            tryCatch(lintr::lint(commandArgs(TRUE), linters = linters, cache = cache),
+            tryCatch(lintr::lint(commandArgs(TRUE), ...),
                     error = function(e) {
                        cat('@@warning: @@', e)
                     })
         }
     }
 };"))
 
+(defun ess-r--find-lintr-file ()
+  "Return the absolute path to the .lintr file.
+Check first the current directory, then the project root, then
+the user's home directory.  Return nil if we couldn't find a .lintr file."
+  (cond (;; current directory
+         (file-readable-p (expand-file-name ".lintr" default-directory))
+         (expand-file-name ".lintr" default-directory))
+        ;; Project root
+        ((and (fboundp 'ess-r-package-project)
+              (ess-r-package-project)) (expand-file-name ".lintr" (cdr (ess-r-package-project))))
+        ((and (fboundp 'project-roots)  ; checking that project-roots is defined can be removed when drop Emacs 24
+              (project-current)
+              (project-roots (project-current)))
```

Comment:
I just discovered that `project-roots` throws on `nil` but `(project-current)` returns nil when three is no project.

@dgutov should this be fixed in project.el?

Reply:
@dgutov `project-roots` doesn't work on custom `('foo . "path/to/project")`. I think the built in method should be defined on arbitrary lists rather on `(head vc)` and `(head transient)`. It would enforce uniform convention.

Reply:
Fixed how?

When there's no current project, your code should decide what to do. You can call `(project-current t)`, of course.

Reply:
Just return nil on nil. If `(project-root (project-current))` is idiomatic way to retrieve roots, then `project-root` should handle the obvious possibility that `project-current` will return nil.